Pattern matching syntax and semantics (#2188)
carbon-language/carbon-lang: proposals/p2188.md

instantiation, patterns that are not meaningful due to a type error are instead
treated as not matching. This includes cases where an `==` fails because of a
missing `EqWith` implementation.

```
fn TypeName[template T:! Type](x: T) -> String {
  match (x) {
    // ✅ OK, the type of `x` is a template parameter.
    case _: i32 => { return "int"; }
    case _: bool => { return "bool"; }
    case _: auto* => { return "pointer"; }
    default => { return "unknown"; }
  }
}
```

Cases where the match is invalid for reasons not involving the template
parameter are rejected when type-checking the template:

```
fn MeaninglessMatch[template T:! Type](x: T*) {
  match (*x) {
    // ✅ OK, `T` could be a tuple.
    case (_: auto, _: auto) => {}
    default => {}
  }
  match (x->y) {
    // ✅ OK, `T.y` could be a tuple.
    case (_: auto, _: auto) => {}
    default => {}
  }
  match (x) {
    // ❌ Error, tuple pattern cannot match value of non-tuple type `T*`.
    case (_: auto, _: auto) => {}
    default => {}
  }
}
```

### Guards

We allow `case`s within a `match` statement to have _guards_. These are not part
of pattern syntax, but instead are specific to `case` syntax:

> -   _case_ ::= `case` _pattern_ [`if` _expression_]? `=>` _block_

Reviewer: Idle and pretty minor comment -- I wonder if we should consider alternative syntaxes for the `=>` part of this... Specifically in case we end up wanting `=>` for any other parts of the grammar.

Not super important though, as this might be the *least* impactful syntax question here, and the easiest for us to change our minds about subsequently. =D

Author: This proposal isn't really considering syntax options for `case` / `match` in general; all this section is trying to describe is where the `if` _expression_ part fits into the existing (but not in an approved design) syntax. We should consider alternatives for `match` and `case` syntax, but not in this proposal.

        not use the `:=` syntax.
    -   If we accept such cases, there is a risk of grammar ambiguities.
    -   If we reject such cases, we may still want to keep `auto` around for
        them, creating inconsistency.
-   Not a complete replacement for `auto` if we want to also allow things like
    `v: Vector(auto)`. C++ doesn't allow the equivalent syntax currently, but it
    was part of the Concepts TS and seems likely to return at some point.
-   No syntactic difference between accidentally omitting a type entirely and
    requesting type deduction. However, the mistake of omitting a type but
    retaining the `:` seems unlikely, and the `:` followed by the absence of a
    type is a signal that something is happening, so this seems to be less of a
    concern than for the `var n = init;` syntax.

See discussion topics
[1](https://github.com/carbon-language/carbon-lang/discussions/1495) and
[2](https://github.com/carbon-language/carbon-lang/discussions/1988).

### Struct pattern syntax

We could omit the `, _` syntax. This would simplify struct patterns, but at the
cost of removing a feature that can be useful for reducing verbosity and making

> library evolution easier.

Reviewer: I'm not fully convinced that we should add this yet TBH.

I wonder if we'll be able to address this usecase *with* variadics, and with tools that also address discarding an unspecified number of tuple elements.

I lean a little bit towards maybe documenting this option but waiting to see if we instead can solve this as part of variadic patterns more generally. What do other folks think here?

The flip side is that I think this is a minor feature and easily *removed* if we find that variadics obviate it, so I also don't think this is a big deal either way.

Reviewer 2: Note that struct fields aren't expressions or values, so the ability to create and match packs of struct fields would be a separate, bespoke feature, not something that would emerge automatically from a more general variadics design. I don't think we've discussed having such a feature, and it's not clear to me if there are any other use cases for it (maybe metaprogramming?). So my inclination would be to keep `, _` for now, and then replace it if and when we discover we want packs-of-struct-fields for other reasons.

Author: I agree that it's not a big deal either way, because this should be easy to change later. I lean towards adding it and seeing if we can replace it with something else later, such as variadics, and similarly seeing if it gets used heavily.

We find that use of pattern-matching in type position can harm readability. For
example, the first of these two examples may be easier to read due to having
less nesting:

```
fn F[T:! Type](x: T);
fn F(x: (T:! Type));
```

Having distinct syntax for type-level matching and value-level matching helps
guide the reader to the correct interpretation, even though the underlying
matching process is expected to be similar or identical. As a result, we keep
type deduction syntax and pattern matching syntax separate for now:

-   In pattern matching, bindings and wildcards are introduced by nested `:` /
    `:!` patterns, and the right-hand side of a binding pattern is never a
    proper pattern.
-   In type deduction, deduced values are specified separately and an expression
    written in terms of those bindings describes the type.

There are some contexts where there is no syntactic location for introducing
deduced values, such as in `case` labels. For syntactic consistency, such cases
should be addressed by adding `forall` syntax, if there is motivation to support
deduction:

```
match (templated_value) {
  case forall [template T:! Type] (p: T*) => { heap.Delete(p); }
}
```

> ### Introducer syntax for expression patterns

Reviewer: I feel like this section doesn't fully capture what we're giving up without this. Is it possible to capture some of the context from the discussion leading up to this Discord message? https://discord.com/channels/655572317891461132/709488742942900284/1020559500597538886

Generally, I wonder if we want to more deeply revisit whether we can and should shift the syntax in this direction. For example, I think I see plausible solutions to the purely syntactic concerns with identifying an introducer below.

That said, I'd be happy to either move that discussion to an issue for leads, and even do that as a fix-forward with even a minor note here that we anticipate revisiting this specific aspect of the syntax to see if there are approaches that might result in an overall better model.

Author: Added some more notes and links. I favor the fix-forward approach, given that the most promising path involves changing the syntax for `where` expressions to remove the other, conflicting meaning of `is`.

`default => {}` case. If we made the opposite choice, we would not automatically
have an easy way to request that the check be performed. We could add syntax to
request it, but such syntax would likely be forgotten and the default behavior
would be that errors are silently permitted. This seems sufficient to outweigh
the potential ergonomic cost of requiring the `default => {}` case to be written
explicitly.

Another motivation for requiring exhaustiveness is that it simplifies other
language rules. For example, when determining whether control flow can reach the
end of a function with a declared return type, a separate exhaustiveness
analysis is not necesasry.

One concern with exhaustiveness checking is that it will cause the addition of
an alternative to a choice type to be a breaking change by default. However,
this is also one of the main advantages, and the design of choice types is
intended to eventually provide a mechanism to specify that a choice type is
extensible, which if used would mean that a set of patterns for that choice type
would only be considered exhaustive if it includes a wildcard pattern.

> ## Future work

Reviewer: Maybe also mention an expression-form as future work?

Author: Done.

```carbon
fn takeDamage(p: player*) {
  match (*p) {
    case {.hitpoints = 0, .lives = 0, _} => { gameOver(); }
    case {.hitpoints = addr hp: i32*, .lives = addr l: i32*, _} if *hp == 0 => {
      *hp = 10;
      --*l;
    }
    case {.hitpoints = addr hp: i32*, _} if *hp <= 3 => {
      --*hp;
      messageAlmostDead();
    }
    case {.hitpoints = addr hp: i32*, _} => {
      --*hp;
    }
  }
}
```

Work in this area will need to consider whether we can provide this feature
ergonomically without introducing reference-like behavior.

### Type deduction

This proposal does not cover type deduction, instead considering it to be a
separate topic from pattern matching syntax, even though the semantic behavior
of the two may be quite similar or identical.

We will need a proposal to explore type deduction and describe its functioning.

## Examples

These examples are translations of examples in WG21 paper
[P0095R1](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0095r1.html)
and
[P1371R3](https://www.open-std.org/JTC1/SC22/WG21/docs/papers/2020/p1371r3.pdf),
with permission from the author of those examples, David Sankel. Thank you,
David!

> ### Examples from P0095R1

Reviewer: Does it make sense to provide a comparison to the syntax proposed in https://wg21.link/p2392 and more generally acknowledge some of the differences in design there?

Author: If that comparison is useful when evaluating this proposal, I can add it, but I'm a little hesitant to get too deep into comparing this with in-flight WG21 proposals. I'm a little tempted to remove the P0095R1 / P1371R3 comparisons, given that they're not comparing against an existing language.

## Problem

Carbon uses patterns wherever a value should be given a name, decomposed, or
matched against: in function parameters, variable declarations, `match`
statement `case`s, `for` statement loop variables, and so on. Simplified forms
of patterns, required to be just a simple name binding, appear in additional
contexts, such as fields in classes and implicit parameter lists. While we have
syntax specified for some of these constructs, we do not have an approved
proposal describing the syntax or semantics of patterns.

## Background

See [pattern matching](https://en.wikipedia.org/wiki/Pattern_matching) on
wikipedia for a broad overview of the subject.

We refer to the value being matched by a pattern as the _scrutinee_.

## Proposal

Patterns in Carbon are a generalization of the expression grammar. Compared to
expressions, patterns add:

-   Bindings, of the form `name: type`, which give a name for the scrutinee.
-   `var` _pattern_, which requests mutable storage be provided for the
    scrutinee.
-   Additional syntax to make matching against structs more convenient.

## Details

### Expressions versus proper patterns

> A pattern that contains within it any pattern-specific syntax, such as a

Reviewer 3: Can you rephrase this?

Author: Done. Better?

Reviewer 3: It is clear.

binding, is a _proper pattern_. Any other pattern must necessarily be an
expression. Many expression forms, such as arbitrary function calls, are not
permitted as proper patterns, so cannot contain bindings.

-   _pattern_ ::= _proper-pattern_

```
fn F(n: i32) -> i32 { return n; }

match (F(42)) {
  // ❌ Error: binding can't appear in a function call.
  case (F(n: i32)) => {}
}
```

### Expression patterns

An expression is a pattern.

-   _pattern_ ::= _expression_

The pattern is compared with the expression using the `==` operator: _pattern_
`==` _scrutinee_.

```
fn F(n: i32) {
  match (n) {
    // ✅ Results in an `n == 5` comparison.
    // OK despite `n` and `5` having different types.
    case 5 => {}
  }
}
```

### Bindings

A name binding is a pattern.

-   _binding-pattern_ ::= `unused`? _identifier_ `:` _expression_

> -   _binding-pattern_ ::= `unused`? _identifier_ `:` _proper-pattern_

Reviewer 4: I feel like this second form is controversial. Is it required?

Author: We could probably replace it with _identifier_ `:` `auto` (and remove `auto` patterns) without affecting many of our current examples. I think this syntax could be valuable in some cases, for example for type-dependent `match`es in templates:

```
fn F[template T:! Type](x: T) -> auto {
  match (x) {
    case (a: (U:! Type), b: (V:! AddableWith(U))) => {
      return F(a + b);
    }
    // ...
  }
}
```

... but if we'd rather not allow this (and maybe permit `case forall` if we find things like the above motivating), I don't think it creates any major problems to remove it. (explorer has only a few testcases using this.)

Reviewer 4: My main arguments against it are readability and consistency with the decision we made for `impl` declarations. How does the alternative read?
```
fn F[template T:! Type](x: T) -> auto {
  match (x) {
    case forall [U:! Type, V:! AddableWith(U)] (a: U, b: V) => {
      return F(a + b);
    }
    // ...
  }
}
```

@mconst

Reviewer 5: I think I don't have a strong preference here.  I do find `case (a: (U:! Type), b: (V:! AddableWith(U)))` hard to read, but I'm not sure `case forall [U:! Type, V:! AddableWith(U)] (a: U, b: V)` is significantly better, especially since it requires the unusual `forall` keyword.  So either way's fine with me.

Author: On discord, @chandlerc seemed pretty strongly in the camp of not permitting patterns in this context, so I'm going with that direction for now. I think the consequence is that we are effectively separating pattern matching from type deduction, and treating them as two different things, and I think that's reasonable. (That's what explorer does, for what it's worth.)

Done.

### Wildcard

A syntax like a binding but with `_` in place of an identifier can be used to
ignore part of a value.

-   _binding-pattern_ ::= `_` `:` _expression_
-   _binding-pattern_ ::= `_` `:` _proper-pattern_

The behavior is equivalent to that of an `unused` binding with a unique name.

```
fn F(n: i32) {
  match (n) {
    // ✅ Matches and discards the value of `n`.
    case _: i32 => {}
    // ❌ Error: unreachable.
    default => {}
  }
}
```

### Generic bindings

A `:!` can be used in place of `:` for a binding that is usable at compile time.

-   _generic-pattern_ ::= `unused`? `template`? _identifier_ `:!` _expression_
-   _generic-pattern_ ::= `template`? `_` `:!` _expression_
-   _proper-pattern_ ::= _generic-pattern_

Note that, unlike in a _binding-pattern_, the type of a generic binding cannot
be a pattern.

```
// ✅ `F` takes a generic type parameter `T` and a parameter `x` of type `T`.
fn F(T:! Type, x: T) {
  var v: T = x;
}
```

The `template` keyword indicates the binding is introducing a template
parameter, so name lookups into the parameter should be deferred until its value
is known.

### `auto`

> The pattern `auto` is shorthand for `_:! Type`.

Reviewer 3: Treat them equivalent?
```
var v: auto = SomeComplicatedExpression(n);
var v: (_:! Type) = SomeComplicatedExpression(n);
```

Author: Added as an example.

Reviewer 3: It looks good.

> The pattern `auto` is shorthand for `_:! Type`.

Reviewer 4: This isn't what I would expect in a function parameter, where I would expect it to be a template.

Author: Typo: I meant for it to be a template parameter in general. Fixed.

-   _proper-pattern_ ::= `auto`

```
fn F(n: i32) {
  var v: auto = SomeComplicatedExpression(n);
}
```

### `var`

A `var` prefix indicates that a pattern provides mutable storage for the
scrutinee.

-   _proper-pattern_ ::= `var` _proper-pattern_

A `var` pattern matches when its nested pattern matches. The type of the storage
is the resolved type of the nested _pattern_. Any bindings within the nested
pattern refer to portions of the corresponding storage rather than to the

> scrutinee, somewhat as if the mutable storage is first initialized from the

Reviewer 3: Can you rephrase this?

Author: Better?

Reviewer 3: lgtm

scrutinee and then the nested pattern is matched against the mutable storage,
but the variable is not actually initialized unless the complete pattern
matches.

```
fn F(p: i32*);
fn G() {
  match ((1, 2)) {
    // `n` is a mutable `i32`.
    case (var n: i32, 1) => { F(&n); }
    // `n` and `m` are the elements of a mutable `(i32, i32)`.
    case var (n: i32, m: i32) => { F(if n then &n else &m); }
  }
}
```

A `var` pattern cannot be nested within another `var` pattern. The declaration
syntax `var` _pattern_ `=` _expresson_ `;` is equivalent to `let` `var`
_pattern_ `=` _expression_ `;`.

### Tuple patterns

A tuple of patterns can be used as a pattern.

-   _tuple-pattern_ ::= `(` [_expression_ `,`]\* _proper-pattern_ [`,`

>     _pattern_]\* `)`

Reviewer 4: Allow optional trailing `,`, which is required to make a 1-tuple?

Author: Trailing `,` for one-tuples is an [open question](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/tuples.md#single-value-tuples). Filed #2191, but I've changed this proposal to allow trailing commas on the assumption that that issue will go the way I think it will :)

Author: #2191 decided: trailing `,` is allowed.

```
match ({.a = 1, .b = 2}) {
  case {.b = n: i32, .a = m: i32} => {}
}
```

The scrutinee is required to be of struct type, and to have the same set of
field names as the pattern. The pattern is matched left-to-right, meaning that
matching is performed in the field order specified in the pattern, not in the
field order of the scrutinee. This is consistent with the behavior of matching
against a tuple-valued expression, where the left operand determines the order
in which `==` comparisons are performed.

In the case where a field will be bound to an identifier with the same name, a
shorthand syntax is available: `a: T` is synonymous with `.a = a: T`.

```
match ({.a = 1, .b = 2}) {
  case {a: i32, b: i32} => { return a + b; }
}
```

If some fields should be ignored when matching, a trailing `, ...` can be added
to specify this:

```
match ({.a = 1, .b = 2}) {
  case {.a = 1, ...} => { return 1; }
  case {b: i32, ...} => { return b; }
}
```

This is valid even if all fields are actually named in the pattern.

### Choice patterns

A choice pattern is used to match one alternative of a choice type.

_proper-pattern_ ::= _callee-expression_ _tuple-pattern_ _proper-pattern_ ::=

> _designator_ _tuple-pattern_?

Reviewer 4: ```suggestion
-   _proper-pattern_ ::= _callee-expression_ _tuple-pattern_
-   _proper-pattern_ ::= _designator_ _tuple-pattern_?
```

alternative that has a parameter list, and the scrutinee is implicitly converted
to that choice type. Otherwise, the scrutinee is required to be of some choice
type, and the designator is looked up in that type and is required to name an
alternative with a parameter list if and only if a _tuple-pattern_ is specified.

The pattern matches if the active alternative in the scrutinee is the specified
alternative, and the arguments of the alternative match the given tuple pattern
(if any).

```
choice Optional(T:! Type) {
  None,
  Some(T)
}

match (Optional(i32).None) {
  // ✅ `.None` resolved to `Optional(i32).None`.
  case .None => {}
  // ✅ `.Some` resolved to `Optional(i32).Some`.
  case .Some(n: i32) => { Print("{0}", n); }
  // ❌ Error, no such alternative exists.
  case .Other => {}
}

class X {
  external impl as ImplicitAs(Optional(i32));
}

match ({} as X) {
  // ✅ OK, but expression pattern.
  case Optional(i32).None => {}
  // ✅ OK, implicitly converts to `Optional(i32)`.
  case Optional(i32).Some(n: i32) => { Print("{0}", n); }
}
```

Note that a pattern of the form `Optional(T).None` is an expression pattern and
is compared using `==`.

> ### Templates

Reviewer 4: We're going to need to have a conversation about whether this should also be allowed for values with checked/symbolic generic types.

Author: Acknowledged. For now I want to reject, given that dispatch based on the actual type seems very template-y.

If the type of the scrutinee of a pattern is either a template parameter or is
an associated type that depends on a template parameter, any checking of the
type of the scrutinee against the type of the pattern is deferred until the
template parameter's value is known. During instantiation, patterns that are not
meaningful due to a type error are instead treated as not matching. This
includes cases where an `==` fails because of a missing `EqWith` implementation.

```
fn TypeName[template T:! Type](x: T) -> String {
  match (x) {
    // ✅ OK, the type of `x` is a template parameter.
    case _: i32 => { return "int"; }
    case _: bool => { return "bool"; }
    case _: auto* => { return "pointer"; }
    default => { return "unknown"; }
  }
}
```

```
fn NeverWorks[template T:! Type](triple: (T, T, T)) {
  match (triple) {
    // ❌ Error: type mismatch matching struct against (T, T, T)
    case {.a: i32, .b: i32} => {}
  }
}
```

> 

Reviewer 3: No dynamic type yet?

Author: Added as future work.

Reviewer 3: it looks good.

        `if constexpr` in many cases.

## Alternatives considered

### Struct pattern syntax

We could omit the `, ...` syntax. This would simplify struct patterns, but at
the cost of removing a feature that can be useful for reducing verbosity and
making library evolution easier.

We could always allow a struct pattern to match a struct with more fields,
without requiring a `, ...` suffix. This would aid evolution by reducing the
cases where adding a field to a struct can be a breaking change, but such cases
would still exist. Further, this would make matches against a struct-valued
expression inconsistent with matches against a struct pattern.

We could remove the `{field: type}` shorthand and require
`{.field = field: type}`. This may avoid encouraging reusing the field name even
when it is not appropriate, and would remove some syntactic sugar that's not
formally necessary. However, we expect this to be a common case whose ergonomics
are important.

We could use a different syntax for `{field: type}` that is less divergent from
other struct syntaxes:

-   `{.field: type}` seems like a contender, but doesn't work because that is

>     already recognized as a tuple type literal. Also, the use of normal binding

Reviewer 4: ```suggestion
    already recognized as a struct type literal. Also, the use of normal binding
```

    syntax means that every locally-introduced name is always introduced by
    `name: type` where `name` is not preceded by `.`.
-   `{.=field: type}` might be a reasonable mnemonic shorthand for
    `{.field = field: type}`, but looks a little surprising. This is probably
    the best choice if concerns are found with `{field: type}` syntax.

> ## Future work

Reviewer 4: Dynamic type tests?

Author: Added.

## Abstract

This paper proposes concrete syntax and semantic choices for Carbon patterns.

## Problem

Carbon uses patterns wherever a value should be given a name, decomposed, or
matched against: in function parameters, variable declarations, `match`
statement `case`s, `for` statement loop variables, and so on. Simplified forms
of patterns, required to be just a simple name binding, appear in additional
contexts, such as fields in classes and implicit parameter lists. While we have
syntax specified for some of these constructs, we do not have an approved
proposal describing the syntax or semantics of patterns.

## Background

See [pattern matching](https://en.wikipedia.org/wiki/Pattern_matching) on
wikipedia for a broad overview of the subject.

We refer to the value being matched by a pattern as the _scrutinee_.

## Proposal

Patterns in Carbon are a generalization of the expression grammar. Compared to
expressions, patterns add:

-   Bindings, of the form `name: type`, which give a name for the scrutinee.
-   `var` _pattern_, which requests mutable storage be provided for the
    scrutinee.
-   Additional syntax to make matching against structs more convenient.

> ## Details

Reviewer 2: This proposal doesn't cover deduced parameters, even as future work, but I've been thinking of them as an integral part of pattern matching. Do you consider them to be separate?

Author: I think the discussion on `name: pattern` seems to have converged on type deduction syntax being separate from pattern matching syntax. The semantic models behind them might be the same, but we don't seem to like using the same syntax for both.

Reviewer 2: My point is more about semantics than syntax, but let's move this discussion to my comment near the bottom.

### Expressions versus proper patterns

Expressions are patterns, as described below. A pattern that is not an
expression, because it contains pattern-specific syntax such as a binding, is a
_proper pattern_. Many expression forms, such as arbitrary function calls, are
not permitted as proper patterns, so cannot contain bindings.

-   _pattern_ ::= _proper-pattern_

```
fn F(n: i32) -> i32 { return n; }

match (F(42)) {
  // ❌ Error: binding can't appear in a function call.
  case (F(n: i32)) => {}
}
```

### Expression patterns

An expression is a pattern.

-   _pattern_ ::= _expression_

The pattern is compared with the expression using the `==` operator: _pattern_
`==` _scrutinee_.

```
fn F(n: i32) {
  match (n) {
    // ✅ Results in an `n == 5` comparison.
    // OK despite `n` and `5` having different types.
    case 5 => {}
  }
}
```

Any `==` operations performed by a pattern match occur in lexical order, but for

> repeated matches against the same `_pattern_`, later comparisons may be skipped

Reviewer 2: ```suggestion
repeated matches against the same _pattern_, later comparisons may be skipped
```

      return other == 1;
    }
  }
}

fn F() {
  // Prints `Matching 1` then `Matching 2`,
  // may or may not then print `Matching 1` again.
  match ((1, 2)) {
    case ({} as ChattyIntMatcher, 0) => {}
    case (1, {} as ChattyIntMatcher) => {}
    case ({} as ChattyIntMatcher, 2) => {}
  }
}
```

### Bindings

A name binding is a pattern.

-   _binding-pattern_ ::= `unused`? _identifier_ `:` _expression_
-   _binding-pattern_ ::= `unused`? _identifier_ `:` _proper-pattern_
-   _proper-pattern_ ::= _binding-pattern_

In the first form, the type of the _identifier_ is specified by the
_expression_. In the second form, the _proper-pattern_ is matched against the
type of the scrutinee, and the result is the type of the _identifier_. In either
case, the scrutinee is implicitly converted to the type of the _identifier_.

```
fn F(p: i32*) {
  match (p) {
    // ✅ Matches `(T:! Type)` against `i32*`.
    // `T` is bound to `i32`, `m` is bound to `p`, and the type of `m` is `i32*`.
    case m: (T:! Type)* => {}
  }
}
```

```
fn G() -> i32 {
  match (5) (
    // ✅ `5` is implicitly converted to `i32`.
    // Returns `5 as i32`.
    case n: i32 => { return n; }
  }
}
```

> The `unused` keyword indicates that the binding is intended to not be used.

Reviewer 2: It might be a good idea to explicitly say that this is intended to restate #2022's decision, not supersede it, just in case we miss some subtle disagreement between them.

Author: Done.

> The `unused` keyword indicates that the binding is intended to not be used.

Reviewer 3: Can you show the use case of  `unused`

Author: There is now an example of this a little later in the proposal, plus a cross-reference to #2022 which introduces this keyword.

tuple-valued expression are matched in the same way because `==` for a tuple
compares fields left-to-right.

### Struct patterns

A struct can be matched with a struct pattern.

-   _proper-pattern_ ::= `{` [_field-init_ `,`]\* _proper-field-pattern_ [`,`
    _field-pattern_]\* `}`
-   _proper-pattern_ ::= `{` [_field-pattern_ `,`]+ `...` `}`
-   _field-init_ ::= _designator_ `=` _expression_
-   _proper-field-pattern_ ::= _designator_ `=` _proper-pattern_
-   _proper-field-pattern_ ::= _binding-pattern_
-   _field-pattern_ ::= _field-init_
-   _field-pattern_ ::= _proper-field-pattern_

A struct pattern resembles a struct literal, with at least one field initialized
with a proper pattern:

```
match ({.a = 1, .b = 2}) {
  // Struct literal as an expression pattern.
  case {.b = 2, .a = 1} => {}
  // Struct pattern.
  case {.b = n: i32, .a = m: i32} => {}
}
```

The scrutinee is required to be of struct type, and to have the same set of
field names as the pattern. The pattern is matched left-to-right, meaning that
matching is performed in the field order specified in the pattern, not in the
field order of the scrutinee. This is consistent with the behavior of matching

> against a tuple-valued expression, where the left operand determines the order

Reviewer 2: ```suggestion
against a struct-valued expression, where the left operand determines the order
```

> against a tuple-valued expression, where the left operand determines the order

Reviewer 2: "_pattern_ operand" might be clearer; as it is, I couldn't remember which operand was on the left in the expression case.

Author: Done.

in which `==` comparisons are performed.

In the case where a field will be bound to an identifier with the same name, a
shorthand syntax is available: `a: T` is synonymous with `.a = a: T`.

```
match ({.a = 1, .b = 2}) {
  case {a: i32, b: i32} => { return a + b; }
}
```

> If some fields should be ignored when matching, a trailing `, ...` can be added

Reviewer 2: I worry that this could be mistaken for a variadic syntax, or perhaps even be ambiguous with one.

As you may recall, the current (very preliminary) plan is to express pack expansion using special operator tokens that start with `...`, such as `...,` and `...{`. I don't anticipate wanting a `...}` operator, so there's no outright ambiguity, but it could easily cause confusion for human readers. Furthermore, we were previously considering using a _trailing_ `...` for those operators, in which case `,...` would be ambiguous between this syntax and the comma pack expansion. I currently think leading `...` is a better choice, but switching back does seem within the realm of possibility.

How about something like `._` as an alternative syntax? It's not perfect, because it looks like it should match exactly one field, but it's the best I've come up with so far.

OTOH I'm not very satisfied with the variadic syntax as a whole, so I'm OK with keeping this struct pattern syntax for now, in the hope that we can come up with a better variadic syntax. I just want to make sure this issue is considered.

Author: I'm finding that I like `, _` more than `, ...` or `, ._`, even if we don't end up using `...` for variadics, so I've switched to that for now.

Reviewer 5: Could we potentially make this match the variadic syntax better, by using something like `..., _`?  That way it looks like a pack expansion of `_`, which seems like a pretty good analogy for what it's doing.

Author: Potentially. That seems like something to consider as part of the variadics proposal... though that syntax seems a bit cumbersome to me, and it might suggest that you could use just `, _` to ignore exactly one field, which I don't think is an interesting facility to provide.

to specify this:

```
match ({.a = 1, .b = 2}) {
  case {.a = 1, ...} => { return 1; }
  case {b: i32, ...} => { return b; }
}
```

This is valid even if all fields are actually named in the pattern.

> ### Choice patterns

Reviewer 2: I'd be inclined to call these "alternative patterns", since they have the form and meaning of a single alternative, not a whole `choice` type.

Author: Done.

cases where adding a field to a struct can be a breaking change, but such cases
would still exist. Further, this would make matches against a struct-valued
expression inconsistent with matches against a struct pattern.

We could remove the `{field: type}` shorthand and require
`{.field = field: type}`. This may avoid encouraging reusing the field name even
when it is not appropriate, and would remove some syntactic sugar that's not
formally necessary. However, we expect this to be a common case whose ergonomics
are important.

We could use a different syntax for `{field: type}` that is less divergent from
other struct syntaxes:

-   `{.field: type}` seems like a contender, but doesn't work because that is
    already recognized as a struct type literal. Also, the use of normal binding
    syntax means that every locally-introduced name is always introduced by
    `name: type` where `name` is not preceded by `.`.
-   `{.=field: type}` might be a reasonable mnemonic shorthand for
    `{.field = field: type}`, but looks a little surprising. This is probably
    the best choice if concerns are found with `{field: type}` syntax.

> ## Future work

Reviewer 2: Deduced parameters?

Author: Added.

### Or patterns

We could provide "or patterns", allowing matching of one pattern or another with
the same handler:

```
match (x) {
  case (m: i32, 0) | (0, m: i32) => { return m; }
}
```

### Guards

We could provide guards for patterns, allowing a pattern to match only if some
predicate involving the bindings holds:

```
match (x) {
  case (m: i32, n: i32) if m + n < 5 => { return m - n; }
}
```

### Matching classes by dynamic type

We could provide a way to match polymorphic class objects based on their dynamic
type. This might be the default when matching a polymorphic class, or might
require opt-in.

### User-defined pattern matching

> We could provide some mechanism for allowing a user-defined type to specify how
> it can be matched by patterns.

Reviewer 2: This is already more or less the plan of record.

```suggestion
We plan to provide a mechanism for allowing a user-defined type to specify how
it can be matched by patterns. See [p0157.md](p0157) for details.
```

    -   [Matching classes with struct patterns](#matching-classes-with-struct-patterns)
    -   [Type deduction](#type-deduction)

<!-- tocstop -->

## Abstract

This paper proposes concrete syntax and semantic choices for Carbon patterns.

## Problem

Carbon uses patterns wherever a value should be given a name, decomposed, or
matched against: in function parameters, variable declarations, `match`
statement `case`s, `for` statement loop variables, and so on. Simplified forms
of patterns, required to be just a simple name binding, appear in additional
contexts, such as fields in classes and implicit parameter lists. While we have
syntax specified for some of these constructs, we do not have an approved
proposal describing the syntax or semantics of patterns.

## Background

See [pattern matching](https://en.wikipedia.org/wiki/Pattern_matching) on
wikipedia for a broad overview of the subject.

We refer to the value being matched by a pattern as the _scrutinee_.

## Proposal

Patterns in Carbon are a generalization of the expression grammar. Compared to
expressions, patterns add:

-   Bindings, of the form `name: type`, which give a name for the scrutinee.

> -   `var` _pattern_, which requests mutable storage be provided for the

Reviewer 6: It is unclear what is meant by mutable storages "provide for" the scrutinee.

Author: Reworded.

-   _pattern_ ::= _expression_

The pattern is compared with the expression using the `==` operator: _pattern_
`==` _scrutinee_.

```
fn F(n: i32) {
  match (n) {
    // ✅ Results in an `n == 5` comparison.
    // OK despite `n` and `5` having different types.
    case 5 => {}
  }
}
```

Any `==` operations performed by a pattern match occur in lexical order, but for
repeated matches against the same _pattern_, later comparisons may be skipped by
reusing the result from an earlier comparison:

```
class ChattyIntMatcher {
  external impl as EqWith(i32) {
    fn Eq[me: ChattyIntMatcher](other: i32) {
      Print("Matching {0}", other);
      return other == 1;
    }
  }
}

fn F() {
  // Prints `Matching 1` then `Matching 2`,
  // may or may not then print `Matching 1` again.
  match ((1, 2)) {
    case ({} as ChattyIntMatcher, 0) => {}
    case (1, {} as ChattyIntMatcher) => {}
    case ({} as ChattyIntMatcher, 2) => {}
  }
}
```

### Bindings

#### Name bindings

A name binding is a pattern.

-   _binding-pattern_ ::= `unused`? _identifier_ `:` _expression_
-   _proper-pattern_ ::= _binding-pattern_

The type of the _identifier_ is specified by the _expression_. The scrutinee is

> implicitly converted to that type.

Reviewer 6: "implicitly converted to that type" implies that there is a copy happening here. Is that what the intent is? It would be good to have examples that take the address of a binding and compare it to the address of the scrutinee. For efficiency I think it is important to guarantee that there are no copies when no conversion is necessary.

Author: I agree that we should have stronger guarantees here. I've added some words and notably some examples.

We don't allow `let` bindings to have their addresses taken, and I don't think we intend to guarantee they have a stable address (or at least, we've not agreed that they do), so I don't think I can easily provide the example you asked for, but I can demonstrate when destructors are run at least.

```
fn F() -> i32 {
  match (5) (
    // ✅ `5` is implicitly converted to `i32`.
    // Returns `5 as i32`.
    case n: i32 => { return n; }
  }
}
```

> As specified in [#2022](p2022.md#the-behavior-of-unused-name-bindings), the

Reviewer 6: This link doesn't seem to work. Maybe use an absolute path instead? Same goes for the other links.

Author: I _think_ this is a consequence of viewing this from a github PR and it would work when checked in, but fixed anyway, just in case.

    // ✅ Matches and discards the value of `n`.
    case _: i32 => {}
    // ❌ Error: unreachable.
    default => {}
  }
}
```

As specified in [#1084](p1084.md), function redeclarations may replace named
bindings with wildcards but may not use different names.

```
fn G(n: i32);
fn H(n: i32);
fn J(n: i32);

// ✅ Does not use `n`.
fn G(_: i32) {}
// ❌ Error: name of parameter does not match declaration.
fn H(m: i32) {}
// ✅ Does not use `n`.
fn J(unused n: i32);
```

#### Generic bindings

A `:!` can be used in place of `:` for a binding that is usable at compile time.

-   _generic-pattern_ ::= `unused`? `template`? _identifier_ `:!` _expression_
-   _generic-pattern_ ::= `template`? `_` `:!` _expression_
-   _proper-pattern_ ::= _generic-pattern_

```
// ✅ `F` takes a generic type parameter `T` and a parameter `x` of type `T`.
fn F(T:! Type, x: T) {
  var v: T = x;
}
```

The `template` keyword indicates the binding is introducing a template
parameter, so name lookups into the parameter should be deferred until its value
is known.

#### `auto` and type deduction

The `auto` keyword is a placeholder for a unique deduced type.

-   _expression_ ::= `auto`

```
fn F(n: i32) {

>   var v: auto = SomeComplicatedExpression(n);

Reviewer 6: Have you considered introducing a shorthand where `auto` can be omitted?

```
var v:= SomeComplicatedExpression(n);
```

Author: #851 considered some such shorthands, but not that one. Added to "Alternatives considered". I think that syntax is probably my favourite so far if we find we want a shorthand for this purpose, but I don't think it's urgent to add a shorter syntax now. We can evolve in this direction in the future if we find a sustained interest in it.

-   _proper-field-pattern_ ::= _binding-pattern_
-   _field-pattern_ ::= _field-init_
-   _field-pattern_ ::= _proper-field-pattern_

A struct pattern resembles a struct literal, with at least one field initialized
with a proper pattern:

```
match ({.a = 1, .b = 2}) {
  // Struct literal as an expression pattern.
  case {.b = 2, .a = 1} => {}
  // Struct pattern.
  case {.b = n: i32, .a = m: i32} => {}
}
```

The scrutinee is required to be of struct type, and to have the same set of
field names as the pattern. The pattern is matched left-to-right, meaning that
matching is performed in the field order specified in the pattern, not in the
field order of the scrutinee. This is consistent with the behavior of matching
against a struct-valued expression, where the expression pattern becomes the
left operand of the `==` and so determines the order in which `==` comparisons
for fields are performed.

In the case where a field will be bound to an identifier with the same name, a
shorthand syntax is available: `a: T` is synonymous with `.a = a: T`.

```
match ({.a = 1, .b = 2}) {
  case {a: i32, b: i32} => { return a + b; }
}
```

> If some fields should be ignored when matching, a trailing `, _` can be added to

Reviewer 6: This is a nice rule.

    already recognized as a struct type literal. Also, the use of normal binding
    syntax means that every locally-introduced name is always introduced by
    `name: type` where `name` is not preceded by `.`.
-   `{.=field: type}` might be a reasonable mnemonic shorthand for
    `{.field = field: type}`, but looks a little surprising. This is probably
    the best choice if concerns are found with `{field: type}` syntax.

### Type pattern matching

We could treat type deduction as a form of pattern matching. For example, we
could allow

```
fn F(a: Vector(T:! Type)) -> T { return a[0]; }
fn G() -> i32 {
  let v: Vector(i32) = (1, 2, 3);
  // Deduces `T = i32`.
  return F(v);
}
```

where the value of `T` is determined by pattern-matching `Vector(T:! Type)`
against the supplied type `Vector(i32)` of `v`. And symmetrically:

```
fn H[m: i32, n: i32](k: i32, (m, n)) { return m + n; }
fn I() {
  // Deduces `m = 2`, `n = 3`.
  H(1, (2, 3));
}
```

This would ensure consistency between pattern matching and deduction,
potentially reducing the number of rules that Carbon developers need to learn.

We find that attempting to unify pattern-matching and type deduction in this way

> harms readability. Having distinct syntax for type-level matching and

Reviewer 6: I don't know that I disagree that the unification harms readability, but your examples are contrived so it's really hard to imagine how this would impact typical code.

Author: Weakened the claim and added an example.

part of a pattern is a specific value. This could be mitigated by permitting
certain kinds of value as patterns without an introducer, such as numeric
literals and `true` and `false`, at the cost of introducing more complexity and
more confusion over which cases require `is` and which do not.

We have also not identified a good choice for the introducer syntax, should we
pursue this direction. `is` is not an ideal choice, because elsewhere in Carbon
syntax, it is a relation between a value and its type, so `is T` may be misread
as matching values whose type is `T`. `==` _expression_ has been suggested, but
that would imply the opposite operand order of that in this proposal --
_scrutinee_ `==` _expression_ rather than _expression_ `==` _scrutinee_ -- which
would compare struct fields in a surprising order that diverges from the order
of comparison for a struct pattern.

## Future work

### Or patterns

We could provide "or patterns", allowing matching of one pattern or another with
the same handler:

```
match (x) {
  case (m: i32, 0) | (0, m: i32) => { return m; }
}
```

### Guards

We could provide guards for patterns, allowing a pattern to match only if some

> predicate involving the bindings holds:

Reviewer 6: A pattern matching syntax without guards is very difficult to use when a guard is needed. It seems to me like this should be part of the base set of operations and its low complexity doesn't seem to warrant an initial exclusion.

Author: OK, I'm sold, added.

```
match (x) {
  case (m: i32, n: i32) if m + n < 5 => { return m - n; }
}
```

### Matching classes by dynamic type

We could provide a way to match polymorphic class objects based on their dynamic
type. This might be the default when matching a polymorphic class, or might

> require opt-in.

Reviewer 6: I think it is important to explore this at least a bit. It isn't clear that the current design can easily support an extension like this.

Author: Added a little exploration with the goal of showing that this can be done within this syntactic model.

### User-defined pattern matching

We plan to provide a mechanism for allowing a user-defined type to specify how
it can be matched by patterns. See [proposal #157](p0157.md) for details.

### Matching classes with struct patterns

> We could allow a class to be matched by a struct patttern that matches its

Reviewer 6: ```suggestion
We could allow a class to be matched by a struct pattern that matches its
```

to that choice type. Otherwise, the scrutinee is required to be of some choice
type, and the designator is looked up in that type and is required to name an
alternative with a parameter list if and only if a _tuple-pattern_ is specified.

The pattern matches if the active alternative in the scrutinee is the specified
alternative, and the arguments of the alternative match the given tuple pattern
(if any).

```
choice Optional(T:! Type) {
  None,
  Some(T)
}

match (Optional(i32).None) {
  // ✅ `.None` resolved to `Optional(i32).None`.
  case .None => {}
  // ✅ `.Some` resolved to `Optional(i32).Some`.
  case .Some(n: i32) => { Print("{0}", n); }
  // ❌ Error, no such alternative exists.
  case .Other => {}
}

class X {
  external impl as ImplicitAs(Optional(i32));
}

match ({} as X) {
  // ✅ OK, but expression pattern.
  case Optional(i32).None => {}
  // ✅ OK, implicitly converts to `Optional(i32)`.
  case Optional(i32).Some(n: i32) => { Print("{0}", n); }
}
```

Note that a pattern of the form `Optional(T).None` is an expression pattern and
is compared using `==`.

### Templates

If the type of the scrutinee of a pattern is either a template parameter or is

> an associated type that depends on a template parameter, any checking of the

Reviewer 4: What about other expressions dependent on a template parameter like `template T:! type, v: Vector(T)`?

Author: My intent was that we would be able to reject a tuple or struct pattern that tries to match `v` when checking the template definition. Hm, maybe that doesn't play well with the phases model for templates: if we don't symbolically evaluate `Vector(T)` because it involves a template parameter, then I guess we can't really ask whether it's a tuple or struct type.

Reviewer 4: Maybe the definition of dependent from #2200 could be used here?

Author: I'm not sure that helps. For example, if we have `a: Vector(T)` and `b: Identity(T)`, I would expect both to have the same properties -- both type expressions are known to be of type `Type`, and for both of them the actual type value depends on a template parameter -- but for `a` we might know something about how to do pattern matching whereas for `b` we do not.

Also, given the extensibility support in https://github.com/carbon-language/carbon-lang/pull/2187, I suppose it might be the case that even for `a` we don't actually know how to do matching until we know `T` -- the `impl as Match` that's used might depend on `T`.

type of the scrutinee against the type of the pattern is deferred until the
template parameter's value is known. During instantiation, patterns that are not
meaningful due to a type error are instead treated as not matching. This
includes cases where an `==` fails because of a missing `EqWith` implementation.

```
fn TypeName[template T:! Type](x: T) -> String {
  match (x) {
    // ✅ OK, the type of `x` is a template parameter.
    case _: i32 => { return "int"; }
    case _: bool => { return "bool"; }
    case _: auto* => { return "pointer"; }
    default => { return "unknown"; }
  }
}
```

```
fn NeverWorks[template T:! Type](triple: (T, T, T)) {
  match (triple) {

>     // ❌ Error: type mismatch matching struct against (T, T, T)

Reviewer 4: Would this error be diagnosed at definition time or instantiation?

Author: My intent was to diagnose at template definition time. But per the previous comment I guess maybe that doesn't fit the model well. Removed.

Reviewer 4: I think we should be able to say that these differences are diagnosed at definition time.

Author: OK, I've added this back. As noted in the previous comment, I don't think the definition of dependent from #2200 is sufficient here, but I think it's sufficient in this design doc to describe the intent -- if it's clear in the template definition that the pattern can never match, then that's an error -- and omit the case analysis with the details.

}
```

As specified in

> [#2022](https://github.com/carbon-language/carbon-lang/pull/2022), the `unused`

Reviewer 2: ```suggestion
As specified in [p2022](p2022.md), the `unused`
```

parameter are rejected when type-checking the template:

```
fn MeaninglessMatch[template T:! Type](x: T*) {
  match (*x) {
    // ✅ OK, `T` could be a tuple.
    case (_: auto, _: auto) => {}
    default => {}
  }
  match (x->y) {
    // ✅ OK, `T.y` could be a tuple.
    case (_: auto, _: auto) => {}
    default => {}
  }
  match (x) {
    // ❌ Error, tuple pattern cannot match value of non-tuple type `T*`.
    case (_: auto, _: auto) => {}
    default => {}
  }
}
```

### Guards

We allow `case`s within a `match` statement to have _guards_. These are not part
of pattern syntax, but instead are specific to `case` syntax:

-   _case_ ::= `case` _pattern_ [`if` _expression_]? `=>` _block_

A guard indicates that a `case` only matches if some predicate holds. The
bindings in the pattern are in scope in the guard:

```
match (x) {
  case (m: i32, n: i32) if m + n < 5 => { return m - n; }
}
```

For consistency, this facility is also available for `default` clauses, so that
`default` remains equivalent to `case _: auto`.

### Refutability, overlap, usefulness, and exhaustiveness

Some definitions:

-   A pattern _P_ is _useful_ in the context of a set of patterns _C_ if _P_ can

>     match any value that no pattern in _C_ matches.

Reviewer 4: The way I read this definition, it implies that if _P_ is useful, then _C union P_ is exhaustive, which I don't think is intended.

```suggestion
-   A pattern _P_ is _useful_ in the context of a set of patterns _C_ if there exists
    a value that _P_ can match that no pattern in _C_ matches.
```

-   A set of patterns _C_ is _exhaustive_ if it matches all possible values.
    Equivalently, _C_ is exhaustive if the pattern `_: auto` is not useful in
    the context of _C_.
-   A pattern _P_ is _refutable_ if there are values that it does not match,

>     that is, if the pattern `_` is useful in the context of {_P_}. Equivalently,

Reviewer 4: ```suggestion
    that is, if the pattern `_: auto` is useful in the context of {_P_}. Equivalently,
```

>     the pattern _P_ is _refuatble_ if the set of patterns {_P_} is not

Reviewer 4: ```suggestion
    the pattern _P_ is _refutable_ if the set of patterns {_P_} is not
```

    exhaustive.
-   A set of patterns _C_ is _overlapping_ if there exists any value that is
    matched by more than one pattern in _C_.

For the purpose of these terms, expression patterns that match a constant tuple,
struct, or choice value are treated as if they were tuple, struct, or
alternative patterns, respectively, and `bool` is treated like a choice type.
Any expression patterns that remain after applying this rule are considered to
match a single value from an infinite set of values so that a set of expression
patterns is never exhaustive:

```
fn IsEven(n: u8) -> bool {
  // Not considered exhaustive.
  match (n) {
    case 0 => { return true; }
    case 1 => { return false; }
    ...
    case 255 => { return false; }
  }
  // Code here is considered to be reachable.
}
```

```
fn IsTrue(b: bool) -> bool {
  match (b) {
    case false => { return false; }
    case true => { return true; }
  }
  // Code here is considered to be unreachable.

> }

Reviewer 4: Might be helpful to explicitly say that in this example the match is considered exhaustive

Author: ```suggestion
fn IsTrue(b: bool) -> bool {
  // Considered exhaustive.
  match (b) {
    case false => { return false; }
    case true => { return true; }
  }
  // Code here is considered to be unreachable.
}
```

          // requiring an irrefutable pattern.
          var 5 = n;
        }
        // ❌ Error, refutable expression pattern `5` used in context
        // requiring an irrefutable pattern.
        fn G(n: i32, 5);
        ```

-   When a set of patterns have no ordering or tie-breaker, it is an error for
    them to overlap unless there is a unique best match for any value that
    matches more than one pattern. However, this situation does not apply to any
    current language rule:

    -   For `match` statements, patterns are matched top-down, so overlap is
        permitted.
    -   We do not yet have an approved design for overloaded functions, but it
        is anticipated that declaration order will be used in that case too.
    -   For a set of `impl`s that match a given `impl` lookup, argument
        deduction is used rather than pattern matching, but overlapping `impl`s
        with the same type structure are an error unless either an `impl` is
        provided to cover the overlap or a `match_first` declaration is used to
        order the `impl`s. (This is a pre-existing rule and is unchanged by this

>         proposal.)

Reviewer 4: I believe the pre-existing rule is stricter: 
```suggestion
    -   For a set of `impl`s that match a given `impl` lookup, argument
        deduction is used rather than pattern matching, but `impl`s with the
        same type structure are an error unless a `match_first` declaration
        is used to order the `impl`s. (This is a pre-existing rule and is
        unchanged by this proposal.)
```

## Rationale

> -   [Software and language evolution](/docs/project/goals.md#software-and-language-evolution)

Reviewer 4: It seems like exhaustiveness is partly about forcing needed changes in the presence of evolution, unless the match statement was explicitly written to handle cases that may be added later.

Author: Done.

### Type pattern matching

We could treat type deduction as a form of pattern matching. For example, we
could allow

```
fn F(a: Vector(T:! Type)) -> T { return a[0]; }
fn G() -> i32 {
  let v: Vector(i32) = (1, 2, 3);
  // Deduces `T = i32`.
  return F(v);
}
```

where the value of `T` is determined by pattern-matching `Vector(T:! Type)`
against the supplied type `Vector(i32)` of `v`. And symmetrically:

```
fn H[m: i32, n: i32](k: i32, (m, n)) { return m + n; }
fn I() {
  // Deduces `m = 2`, `n = 3`.
  H(1, (2, 3));
}
```

This would ensure consistency between pattern matching and deduction,
potentially reducing the number of rules that Carbon developers need to learn.

We find that attempting to unify pattern-matching and type deduction in this way
harms readability. Having distinct syntax for type-level matching and
value-level matching helps guide the reader to the correct interpretation, even
though the underlying matching process is expected to be similar or identical.
As a result, we keep type deduction syntax and pattern matching syntax separate
for now:

-   In pattern matching, bindings and wildcards are introduced by nested `:` /

>     `:!` patterns.

Reviewer 2: ```suggestion
-   In pattern matching, bindings and wildcards are introduced by nested `:` /
    `:!` patterns, and the right-hand side of a binding pattern is never a proper
    pattern.
```

convertible to the class type. However, without a design for user-defined
pattern matching and matching on dynamic type, there is a significant risk that
this would conflict with the rules there, so it is deferred for now.

### Type deduction

This proposal does not cover type deduction, instead considering it to be a
separate topic from pattern matching syntax, even though the semantic behavior

> of the two may be quite similar or identical.

Reviewer 2: This is a proposal for pattern matching syntax _and_ semantics, though. Even at the syntactic level, they're closely connected: a deduced parameter list always comes immediately before a pattern, and always consists of a list of binding patterns. At the semantic level, it's not just that the behavior of the two is "similar", it's that they are directly coupled: as far as I can tell, type deduction is always driven by a pattern matching operation, and I think it can feed back into that pattern matching operation.

To be clear, I'm not saying this proposal needs to have a design for type deduction; I just don't think type deduction is as cleanly separable from pattern matching as this makes it sound.

Author: I don't think they're *necessarily* coupled. I think we are confident that we're going to do deduction when performing function calls to generic functions and when selecting an `impl`. I think we are confident that we're going to do pattern matching in a `match` statement and in a `let` or `var` binding. I don't think it's guaranteed that we'll do pattern matching in a function call -- it may be reasonable to restrict a function to have only a list of bindings as its parameters, which only needs pattern matching in a degenerate sense, and I think we're not going to do pattern matching in `impl` selection. So I think there might actually be no situations where we perform both type deduction and non-degenerate pattern matching.

I also think it's possible that we'll model the matching process for the two cases quite differently: for pattern matching, we have a top-down process where the pattern syntax determines what happens at each level (such as evaluating an expression pattern and performing an `==` comparison, or converting the scrutinee), whereas for type deduction I think it's likely we'll instead perform a symbolic evaluation of the type expression (the "pattern") and perform a value comparison between (symbolic) type values to find a set of substitutions. The difference is fairly subtle, but I think deduction ends up having stricter matching semantics than a pattern match -- identical after substitution versus a looser notion of "match".

Reviewer 2: > So I think there might actually be no situations where we perform both type deduction and non-degenerate pattern matching.

It seems like variadic functions will be such a situation. Maybe we can say that semantically, variadics relate only to deduction, and not pattern matching (although I'm not sure of that), but syntactically they're going to appear in the parentheses as well as the square brackets.

This would ensure consistency between pattern matching and deduction,
potentially reducing the number of rules that Carbon developers need to learn.

We find that use of pattern-matching in type position can harm readability. For
example, the first of these two examples may be easier to read due to having
less nesting:

```
fn F[T:! Type](x: T);
fn F(x: (T:! Type));
```

Having distinct syntax for type-level matching and value-level matching helps
guide the reader to the correct interpretation, even though the underlying
matching process is expected to be similar or identical. As a result, we keep

> type deduction syntax and pattern matching syntax separate for now:

Reviewer 4: Address contexts where you can't write `[...]` like `match` statements?

Author: Done.

would compare struct fields in a surprising order that diverges from the order
of comparison for a struct pattern.

### Allow guards on arbitrary patterns

We could treat guards as part of pattern syntax instead of as part of `case`
syntax. However, since guards make a pattern refutable, this wouldn't allow them
anywhere other than in `case`s in the current language design. It would allow
them to be nested within cases:

```
match (x) {
  case (n: i32 if n > 5, "some string") => { ... }
}
```

Such nesting might allow an expensive later check to be avoided. For example, in
the above case we can avoid an `==` comparison on a string if a cheaper
comparison of `n > 5` fails. However, this would introduce complexity into the
grammar, and it's not clear that this feature would add sufficient value to
justify that complexity.

> An additional concern is that if we add `let`...`else` syntax, this would

Reviewer 4: Add a reference to this discussion? https://discord.com/channels/655572317891461132/748959784815951963/981691016040030279

Author: Added.

if each pattern has a constant value and between those constant values, all
possible values of the type are covered. The advantage of this would be that we
improve the precision of our language rules.

This change in rules has some disadvantages and problems:

-   It would add some complexity to the rules and to implementations in order to
    track whether all possible values have been created.
-   For even the simplest types where this would apply, such as `i8`, it seems
    unlikely that a `match` covering all possible values would be written, due
    to the large number of patterns required.
-   In many cases, the value being matched will carry an invariant so that
    matching a subset of the representable values would match all meaningful
    values. We would still have imprecise rules in those cases.
-   Expression patterns are matched with `==`, and for an arbitrary `==` it is
    not computable in general to tell whether a given set of values is
    exhaustive, so we would only be able to apply this in some subset of cases.
-   This restriction is straightforward to work around by replacing the final

>     value match with a `default` or `_`.

Reviewer 4: It seems less error-prone to add a `default` than replace a match with a `default`

Author: Added that as a suggested option.

    return stream << "Rotate by " << r.degrees << " degrees.\n"
  }
};
}

std::ostream& operator<<(std::ostream& stream, const command& cmd) {
  return std::visit(std::bind(Output(), std::ref(stream), std::placeholders::_1),
                    cmd.value);
}
```

</td><td>

```c++
std::ostream& operator<<(std::ostream& stream, const command& cmd) {
  return inspect(cmd) {
    set_score value =>
      stream << "Set the score to " << value << ".\n"
    fire_missile _ =>
      stream << "Fire a missile.\n"
    fire_laser intensity =>
      stream << "Fire a laser with " << intensity << " intensity.\n"
    rotate degrees =>
      stream << "Rotate by " << degrees << " degrees.\n"
  }
}
```

</td><td>

```carbon
impl command as Printable {
  fn Print[me: Self](stream: Cpp.std.ostream*) {
    match (me) {
      case .set_score(value: u64) => {
        stream->Print("Set the score to {0}.\n", value);
      }
      case .fire_missile => {
        stream->Print("Fire a missile.\n");
      }
      case .fire_laser(intensity: u32) => {
        stream->Print("Fire a laser with {0} intensity.\n", intensity);
      }
      case .rotate(degrees: f64) => {

>         stream->Print("Rotate by {9} degrees.\n", degrees);

Reviewer 4: ```suggestion
        stream->Print("Rotate by {0} degrees.\n", degrees);
```
